### ISSUE TITLE
Inconsistent promotion of `this` in extension methods

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#44652]
The following code is accepted by the CFE but rejected by the analyzer:
```dart
class C {
  int? cProp;
}

extension on C? {
  void testCQuestion() {
    if (this != null) {
      print(this.cProp); // (1)
      print(cProp); // (2)
    }
  }
}

class D extends C {
  int? dProp;
}

extension on C {
  void testC() {
    if (this is D) {
      print(this.dProp); // (3)
      print(dProp); // (4)
    }
  }
}

main() {
  C().testCQuestion();
  C().testC();
  D().testCQuestion();
  D().testC();
  (null as C?).testCQuestion();
}
```
The analyzer considers the lines marked (1), (2), (3), and (4) to be erroneous because it considers `this` to be exempt from type promotion (since it is not a variable).  The CFE, on the other hand, treats `this` like a variable inside extension methods, so it allows it to be promoted.

We need to figure out which behavior we want to allow, and make the analyzer and CFE consistent.

**Comment by stereotype441** [comment]
My recollection is that we discussed this a long time ago (~1 year ago perhaps) and decided that the behavior we want is for `this` *not* to be promotable.  @scheglov, @johnniwinther, @leafpetersen can any of you confirm my memory here?

For now I'm going to assume my memory is correct and classify this as a CFE bug.  If others remember it differently I'm happy to discuss.

I'm currently working in an area that is affected by this, so I'm going to look and see if I can find an easy fix.

**Comment by scheglov** [comment]
Yes, I also remember that we decided that `this` should not be promotable.

**Comment by stereotype441** [comment]
Note: the same situation applies in legacy mode, i.e. this code is accepted by the CFE but not the anlayzer:
```dart
// @dart=2.9

class C {
  int cProp;
}

class D extends C {
  int dProp;
}

extension on C {
  void testC() {
    if (this is D) {
      print(this.dProp); // (1)
      print(dProp); // (2)
    }
  }
}

main() {
  C().testC();
  D().testC();
}
```
I believe I can fix this as part of the same CL, but it makes the fix a little less safe.  *Technically* it's a breaking change, but in practice I think it's unlikely to affect anyone because the analyzer has never allowed `this` in extensions to promote, so anyone who would be broken by this change is not analyzing their code using the analyzer.

@leafpetersen WDYT?  Should I go ahead with a fix to both legacy and NNBD behavior?  Should I fix one and not the other?  Should we do a breaking change announcement?

**Comment by stereotype441** [comment]
The fix looks pretty easy.  I prototyped two variants:
- Fix NNBD only: https://dart-review.googlesource.com/c/sdk/+/178920
- Fix both NNBD and legacy: https://dart-review.googlesource.com/c/sdk/+/178922

It looks like there are no existing language tests covering this behavior, so I'm going to write some language tests before sending either CL out for review.  Also, before landing, I'll verify that no internal code would be broken by this change.

**Comment by stereotype441** [comment]
> The fix looks pretty easy. I prototyped two variants:
> 
> * Fix NNBD only: https://dart-review.googlesource.com/c/sdk/+/178920
> * Fix both NNBD and legacy: https://dart-review.googlesource.com/c/sdk/+/178922
> 
> It looks like there are no existing language tests covering this behavior, so I'm going to write some language tests before sending either CL out for review. Also, before landing, I'll verify that no internal code would be broken by this change.

Ok, I've got language tests and both CLs are green on trybots.  I'm going to send both to @johnniwinther for review now so that he can look at them during European daytime hours; hopefully tomorrow US time @leafpetersen and I can figure out which CL (if any) we want to actually land.

**Comment by stereotype441** [comment]
In the language team meeting, we agreed that we want to fix both NNBD and legacy behaviors.  However, since fixing the legacy behavior is technically a breaking change, it needs to go through the breaking change process.

I have some other work that depends on the NNBD fix, so I'm going to go ahead and land that now; then I'll start the breaking change process and hopefully in a few days I'll be able to land the legacy fix.
